dask_match/expr.py

```diff
@@ -1124,12 +1185,13 @@ def _divisions(self):
     def _task(self, index):
         graph = {self._name: (self.exprs[0]._name, index)}
         for _expr in self.exprs:
+            task_index = _expr._partitions[index]
```

Comment:
Just calling out here that it feels off that `_partitions` has to affect code in `Fused`.  This seems like an abstraction that leaks around to others, and leads to a hard-to-understand codebase.  I don't see a better way currently, but I want to call this out as concerning so that we don't get too comfortable doing things like this.

Reply:
Could we handle this in the IO subclasses instead with something like the following?

```python
index 21bc986..258cfa2 100644
--- a/dask_match/expr.py
+++ b/dask_match/expr.py
@@ -1185,13 +1185,12 @@ class Fused(Blockwise):
     def _task(self, index):
         graph = {self._name: (self.exprs[0]._name, index)}
         for _expr in self.exprs:
-            task_index = _expr._partitions[index]
             if isinstance(_expr, Fused):
-                (_, subgraph, name) = _expr._task(task_index)
+                (_, subgraph, name) = _expr._task(index)
                 graph.update(subgraph)
                 graph[(name, index)] = name
             else:
-                graph[(_expr._name, index)] = _expr._task(task_index)
+                graph[(_expr._name, index)] = _expr._task(index)

         for i, dep in enumerate(self.dependencies()):
             graph[self._blockwise_arg(dep, index)] = "_" + str(i)
diff --git a/dask_match/io/csv.py b/dask_match/io/csv.py
index 2f139ff..b7cd441 100644
--- a/dask_match/io/csv.py
+++ b/dask_match/io/csv.py
@@ -35,4 +35,6 @@ class ReadCSV(BlockwiseIO):
         return next(iter(dsk.values()))[0]

     def _task(self, index: int):
+        if self._partitions is not None:
+            index = self._partitions[index]
         return (self._io_func, self._tasks[index][1])
diff --git a/dask_match/io/io.py b/dask_match/io/io.py
index fdd9ba5..c4e8e2b 100644
--- a/dask_match/io/io.py
+++ b/dask_match/io/io.py
@@ -75,6 +75,8 @@ class FromPandas(BlockwiseIO):
         return self._divisions_and_locations[1]

     def _task(self, index: int):
+        if self._partitions is not None:
+            index = self._partitions[index]
         start, stop = self._locations()[index : index + 2]
         return self.frame.iloc[start:stop]

diff --git a/dask_match/io/parquet.py b/dask_match/io/parquet.py
index aac3bd2..989f670 100644
--- a/dask_match/io/parquet.py
+++ b/dask_match/io/parquet.py
@@ -291,4 +291,6 @@ class ReadParquet(BlockwiseIO):
         return self._plan["divisions"]

     def _task(self, index: int):
+        if self._partitions is not None:
+            index = self._partitions[index]
         return (self._plan["func"], self._plan["parts"][index])
```

This doesn't work, but I haven't investigated deeply yet.

Reply:
Right - This PR originally proposed that any `Expr` supporting `_partitions` would need to account for this operand in its `_task` implementation. When `_layer` was agnostic to `_partitions`, there were no problems with `Fused`.  I only attempted to move the logic into `_layer` in response to [your comment](https://github.com/mrocklin/dask-match/pull/42#discussion_r1170612271), so I'd be happy to go in a different direction.

Reply:
Took a quick stab at the `PartitionsFiltered`, and moved the logic out of `_layer`. Let me know if you think that direction makes sense.

Reply:
Good evidence of the fact that I am often wrong!